crypto/async/async_wait.c

```diff
@@ -148,6 +148,33 @@ int ASYNC_WAIT_CTX_clear_fd(ASYNC_WAIT_CTX *ctx, const void *key)
             continue;
         }
         if (curr->key == key) {
+            /* If fd has just been added, remove it from the list */
+            if (curr->add == 1) {
+                if (ctx->fds == curr) {
+                    ctx->fds = curr->next;
+                }
+                else {
+                    struct fd_lookup_st *prev = ctx->fds;
+                    while (prev->next != curr && prev->next != NULL) {
+                        prev = prev->next;
+                    }
+                    if (prev->next == NULL) {
+                        return 1;
+                    }
+                    prev->next = curr->next;
+                }
+
+                /*
+                 * The fd has just been added so it can't be used externally
+                 * and it is safe to call the cleanup function here
+                 */
+                if (curr->cleanup != NULL)
+                    curr->cleanup(ctx, curr->key, curr->fd, curr->custom_data);
```

Comment:
This is a *corner case* where the engine has already added the fd to the wait context but following a failure it has to remove it. This is done before exiting the async job and returning the control to the application.

In a normal scenario the cleanup function won't be called, matching the behavior explained in the documentation.

The alternative would be to cleanup the file descriptor in the engine before calling this function, since this is the last chance to close the fd before freeing the memory.